Sprint 12

DanteIT94 opened the pull request:
1) First Commit - Trying to realise ImagesListService(get-request)

DanteIT94 committed:
Merge branch 'sprint_11'

DanteIT94 committed:
Sprint 12 - First Commit - Trying to create ImagesListService

DanteIT94 committed:
1) ImageListService - for updating list of Pictures

DanteIT94 committed:
Adding LikeService with UI interactions

DanteIT94 committed:
SingleImageViewController -> Now u can see FullImage ^)

DanteIT94 committed:
Logout func was added

DanteIT94 committed:
Updating DataFormatter + Adding Alert on LogoutButton

DanteIT94 committed:
Stylistic updates of files - adding marks

DanteIT94 committed:
Updated OAuthService + was added GradientLayer for Profile

DanteIT94 committed:
Was added animatedGradientLayer for ImageListCells

ABashkirova commented on pinterestClone.xcodeproj/project.pbxproj:
```diff
@@ -32,7 +32,11 @@
 		92F5904A2A0CD259002727E3 /* TabBarViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 92F590492A0CD259002727E3 /* TabBarViewController.swift */; };
 		92F5904C2A111661002727E3 /* OAuthTokenResponseBody.swift in Sources */ = {isa = PBXBuildFile; fileRef = 92F5904B2A111661002727E3 /* OAuthTokenResponseBody.swift */; };
-		92F5904E2A12CAE8002727E3 /* KeyChainManager.swift in Sources */ = {isa = PBXBuildFile; fileRef = 92F5904D2A12CAE8002727E3 /* KeyChainManager.swift */; };
+		92F5904E2A12CAE8002727E3 /* KeyChainManager(NEED TO ADD).swift in Sources */ = {isa = PBXBuildFile; fileRef = 92F5904D2A12CAE8002727E3 /* KeyChainManager(NEED TO ADD).swift */; };
```
Подскажи, в чем потребность указывать в скобочках такие штуки как `(NEED TO ADD)`, `JSON` и так далее?

я не видела такой практики в нейминге, и в целом стараются избегать в названиях такие символы и смешение стилей (капс, camelCase)

DanteIT94 replied:
Ставил себе зрительные напоминания (чтобы не забыть вернуться к ним при оптимизации / рефакторинг кода)

ABashkirova commented on pinterestClone/ImagesListFolder/ImagesListService.swift:
```diff
@@ -0,0 +1,125 @@
+            case .failure(let error):
+                print("Failed to fetch photos:", error)
+                task = nil
+                return
```
return тут не нужен, выход из тела кложуры будет сделан естественным образом по завершению всех инструкций

DanteIT94 replied:
Когда писал эту часть кода, видимо разум уже затуманился а потому просто пропускал его

ABashkirova commented on pinterestClone/ImagesListFolder/ImagesListService.swift:
```diff
@@ -0,0 +1,125 @@
+    private var task: URLSessionTask?
+    private let urlSession = URLSession.shared
+    private let dateFormatter = ISO8601DateFormatter()
+    
+    
+    
```
давай договоримся не оставлять больше одной разделительной пустой строки

DanteIT94 replied:
Понял, исправлюсь!

ABashkirova commented on pinterestClone/ImagesListFolder/ImagesListService.swift:
```diff
@@ -0,0 +1,125 @@
+        guard let token = tokenStorage.token else {return}
+        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
+        
+        let dataTask = urlSession.objectTask(for: request) {[weak self] (result: Result<[PhotoResult],Error>) in
```
фигурные скобки, перечисления сопровождаются пробелами, чтобы код не слипался 
```suggestion
        let dataTask = urlSession.objectTask(for: request) { [weak self] (result: Result<[PhotoResult], Error>) in
```

DanteIT94 replied:
Ок

ABashkirova commented on pinterestClone/ImagesListFolder/ImagesListViewController.swift:
```diff
@@ -15,35 +17,41 @@ final class ImagesListViewController: UIViewController {
+            forName: ImagesListService.DidChangeNotification,
+            object: nil,
+            queue: .main) { [weak self] _ in
+                guard let self else {return}
```
```suggestion
                guard let self else { return }
```

DanteIT94 replied:
Ок

ABashkirova commented on pinterestClone/ImagesListFolder/ImagesListViewController.swift:
```diff
@@ -112,20 +139,77 @@ extension ImagesListViewController: UITableViewDataSource {
+            guard let self = self else {return}
+            switch result {
+            case .success:
+                DispatchQueue.main.async {
+                    ///синх. массив картинок с сервисом
+                    self.photos = self.imageListService.photos
+                    ///Измен
+                    cell.setIsLiked(self.photos[indexPath.row].isLiked)
+                    UIBlockingProgressHUD.dismiss()
+                }
+            case .failure:
+                UIBlockingProgressHUD.dismiss()
+                self.showAlertViewController()
```
очень странно, что в блоке успеха ты переводишь на главную очередь, а в ошибке нет. Выглядит как неконсистенность: надо выяснить нужно ли обернуть в DispatchQueue.main.async код из ошибки, или наоборот убрать обертку в успехе.

DanteIT94 replied:
Так, почитал статьи и немного разобрался -> мы в первую очередь очередь на главный поток пускаем UI-логику чтобы пользователю все отображалось все корректно

По такой логике действительно и case .failure должен обрабатываться в main

ABashkirova commented on pinterestClone/ImagesListFolder/ImagesListViewController.swift:
```diff
@@ -112,20 +139,77 @@ extension ImagesListViewController: UITableViewDataSource {
+
+//MARK: -DateFormatter
+extension ImagesListViewController {
+    private  var dateFormatter: DateFormatter {
+        let formatter = DateFormatter()
+        formatter.dateFormat = "dd MMMM yyyy"
+        formatter.locale = Locale(identifier: "ru")
+        formatter.dateStyle = .medium
+        formatter.timeStyle = .none
+        return formatter
+    }
```
это вычисляемое свойство, при обращении к нему ты будешь каждый раз создавать новый форматтер. Это не оптимально, и в чеклисте требовалось создавать форматтер всего раз. Связанно это с достаточно большими затратами на создание и настройку форматтера (https://sarunw.com/posts/how-expensive-is-dateformatter), а обращение к этому свойству при формировании ячейки, вполне может вызывать тормоза на слабых девайсах.

DanteIT94 replied:
Александра, добрый день!  Спасибо за важное уточнение! Со статьей ознакомился, однако некоторые моменты пока что прошли "мимо меня" так как специфику загруженности и "дороговизны" кода я наверно во всей красе пойму на реальных проектах(

В итоге решил пойти по пути Первого проекта -> вынес форматтер в отдельный файл + добился удаления "г." после даты в целях  единообразия проекта и макета.

ABashkirova commented on pinterestClone/UserProfileFolder/ProfileViewController.swift:
```diff
@@ -135,4 +155,153 @@ final class ProfileViewController: UIViewController {
+    
+    //MARK: -Логаут из акка
+    private func accountLogout() {
+        tokenStorage.keychainWrapper.removeObject(forKey: "token")
```
.keychainWrapper - это зависимость стораджа, должна быть закрыта приватным модификатором. Стучаться через нее чтобы удалить свойство - архитектурная ошибка (не соблюдение принципов сокрытия реализации).

DanteIT94 replied:
Действительно, видимо я просто не привык работать еще кейчейнами - просто напрямую вызвал KeychainWrapper.standart в методе чтобы не скрещиваться с storage)
Спасибо за наглядное выделение архитектурных ошибок! С каждым днем буду становиться лучше)

ABashkirova reviewed (changes requested):
К сожалению работу проверить не удалось, после авторизации дальше не пускает. 

скорее всего проблема с моделью профиля: поле lastName может приходить опциональным. Это можно проверить, изменив настройки профиля - сохранить все поля пустые, например: 

<img width="751" alt="Снимок экрана 2023-05-09 в 21 07 21" src="https://github.com/DanteIT94/pinterestClone/assets/20269264/d9911b83-809d-45a5-b794-e5052bd211d3">

Без исправления этой ошибки я не смогу проверить твою работу

DanteIT94 committed:
Working on review # 1

1) Was corrected errors with users account (in case someone dont have surname)

2) The formatter was moved to a separate file (and an extension for the Date type was added)

3) Now KeyChainStorage and deleting a token dont cross

4) Added minor stylistic changes

IBDesignable commented on pinterestClone/UserProfileFolder/Models/ProfileResultModel.swift:
```diff
@@ -0,0 +1,25 @@
+    let lastName: String?
+    let bio: String?
+    
+    ///Определяем свойства структуры, которые соответствуют полям ответа сервера.
+    enum CodingKeys: String, CodingKey {
+        case username = "username"
+        case firstName = "first_name"
+        case lastName = "last_name"
+        case bio = "bio"
```
CodingKeys – это один из вариантов решения задачи. Также можно настроить JsonDecoder таким образом, чтобы он данные сам преобразовывал из snake_case до camelCase

https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable  - можно подробнее тут

DanteIT94 replied:
Ого! Даже не знал о таком функционале - попробую имплементировать!

IBDesignable commented on pinterestClone/UserProfileFolder/ProfileViewController.swift:
```diff
@@ -135,4 +152,151 @@ final class ProfileViewController: UIViewController {
+        let alert = UIAlertController(title: "Пока, Пока!", message: "Уверены, что хотите выйти?", preferredStyle: .alert)
+        
+        let yesAction = UIAlertAction(title: "Да", style: .default) { [weak self] _ in
+            guard let self = self else {return}
```
```suggestion
            guard let self = self else { return }
```

IBDesignable commented on pinterestClone/UserProfileFolder/ProfileViewController.swift:
```diff
@@ -135,4 +152,151 @@ final class ProfileViewController: UIViewController {
+    
+    //MARK: -Логаут из акка
+    private func accountLogout() {
+        KeychainWrapper.standard.removeObject(forKey: "token")
```
Может в прошлый раз выразилась некорректно, но в предыдущей версии был TokenStorage, который работал сначала с UserDefaults, так и почему бы им дальше и не пользоваться: нужно прятать внешние зависимости, оборачивая их в интерфейс на языке приложения.
например `resetToken`.

DanteIT94 replied:
Ааа, понял! Прячем внешние зависимости, чтобы в случае их "упразднения" можно было бы только в одном месте отредактировать код! Исправлю!

DanteIT94 replied:
Готово!

IBDesignable commented on pinterestClone/UserProfileFolder/ProfileViewController.swift:
```diff
@@ -135,4 +152,151 @@ final class ProfileViewController: UIViewController {
+
+//MARK: Наводим градиентную анимацию
+extension ProfileViewController {
+    private func setAvatarGradient() {
```
круто что сделал анимацию 👍🏻

DanteIT94 replied:
Дальше - Больше!!

IBDesignable commented on pinterestClone/ImagesListFolder/SingleImageFolder/SingleImageViewController.swift:
```diff
@@ -120,8 +143,26 @@ final class SingleImageViewController: UIViewController {
     }
     
+    private func showError() {
+        let alertVC = UIAlertController(
+            title: "Что-то пошло не так(",
+            message: "Повторим еще раз?",
+            preferredStyle: .alert)
+        let actionNo = UIAlertAction(title: "Не надо", style: .default) {[weak self] _ in
+            self?.dismiss(animated: true)
+        }
+        let actionAgain = UIAlertAction(title: "Повторить", style: .default) { [weak self] _ in
+            self?.fetchFullImage()
+        }
+        
+        alertVC.addAction(actionNo)
+        alertVC.addAction(actionAgain)
+        present(alertVC, animated: true)
```
Задача-челендж: написать AlertPresenter, аналогичный тому из MovieQuiz, чтобы работать с ним в этом приложении

DanteIT94 replied:
Челендж принят!

IBDesignable commented on pinterestClone/ImagesListFolder/SingleImageFolder/SingleImageViewController.swift:
```diff
@@ -88,6 +93,24 @@ final class SingleImageViewController: UIViewController {
+            case .success(let ResultImage):
+                createSingleImageView()
+                self.rescaleAndCenterImageInScrollView(image: ResultImage.image)
+            case .failure(_):
```
Можно упростить
```suggestion
            case .failure:
```

IBDesignable commented on pinterestClone/ImagesListFolder/SingleImageFolder/SingleImageViewController.swift:
```diff
@@ -88,6 +93,24 @@ final class SingleImageViewController: UIViewController {
+            guard let self = self else {return}
+            UIBlockingProgressHUD.dismiss()
+            switch result {
+            case .success(let ResultImage):
+                createSingleImageView()
+                self.rescaleAndCenterImageInScrollView(image: ResultImage.image)
```
последи за camelCase в своем коде:
```suggestion
            case .success(let resultImage):
                createSingleImageView()
                self.rescaleAndCenterImageInScrollView(image: resultImage.image)
```

IBDesignable commented on pinterestClone/ImagesListFolder/SingleImageFolder/SingleImageViewController.swift:
```diff
@@ -88,6 +93,24 @@ final class SingleImageViewController: UIViewController {
     }
     
+    func fetchFullImage() {
+        guard let fullImageUrl = fullImageUrl else {return}
```
и еще разок :)
```suggestion
        guard let fullImageUrl = fullImageUrl else { return }
```

IBDesignable commented on pinterestClone/ImagesListFolder/SingleImageFolder/SingleImageViewController.swift:
```diff
@@ -88,6 +93,24 @@ final class SingleImageViewController: UIViewController {
+    func fetchFullImage() {
+        guard let fullImageUrl = fullImageUrl else {return}
+        UIBlockingProgressHUD.show()
+        imageView.kf.setImage(with: fullImageUrl) {[weak self] result in
```
```suggestion
        imageView.kf.setImage(with: fullImageUrl) { [weak self] result in
```

IBDesignable commented on pinterestClone/ImagesListFolder/ImagesListViewController.swift:
```diff
@@ -86,15 +105,22 @@ extension ImagesListViewController: UITableViewDelegate {
         return cellHeight
     }
+    
+    ///В этом методе вызываем метод fetchPhotosNextPage из ImagesListService
```
после док-коммента пробел оставляют:
```suggestion
    /// В этом методе вызываем метод fetchPhotosNextPage из ImagesListService
```

IBDesignable commented on pinterestClone/Helpers/Date+Extension .swift:
```diff
@@ -0,0 +1,24 @@
+
+import Foundation
+
+private  let dateFormatter: DateFormatter = {
+    let formatter = DateFormatter()
+    formatter.dateFormat = "dd MMMM yyyy"
+    formatter.dateStyle = .medium
+    formatter.timeStyle = .none
+    return formatter
+} ()
+
+extension Date {
+    var dateTimeString: String {
+            var dateString = dateFormatter.string(from: self)
+            dateString = dateString.replacingOccurrences(of: "г.", with: "")
+            return dateString
+        }
+}
```
давай отступ поправим : control + i

DanteIT94 committed:
was changed method of json-decoding (throught . keyDecodingStrategy)
